benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java

```diff
@@ -150,52 +158,145 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
+                    return queryExecutionResults;
+                });
+            }
+        }
+        return executionCallables;
+    }
 
-                        if (reportStatus) {
-                            statusReporter.reportExecutionStarted(queryExecution);
-                        }
-                        QueryExecutionResultBuilder failureResult = new QueryExecutionResultBuilder(queryExecution)
-                                .startTimer();
-                        try {
-                            result = queryExecutionDriver.execute(queryExecution, connection);
-                        }
-                        catch (Exception e) {
-                            LOG.error("Query Execution failed for benchmark {}", benchmark.getName());
-                            result = failureResult
-                                    .endTimer()
-                                    .failed(e)
-                                    .build();
-                        }
+    private List<Callable<List<QueryExecutionResult>>> buildConcurrencyQueryExecutionCallables(Benchmark benchmark, int runs, boolean reportStatus, Optional<ZonedDateTime> lastTime)
+    {
+        List<Callable<List<QueryExecutionResult>>> executionCallables = newArrayList();
+        for (int thread = 0; thread < benchmark.getConcurrency(); thread++) {
+            int finalThread = thread;
+            executionCallables.add(() -> {
+                LOG.info("Running throughput test: {} queries, {} runs", benchmark.getQueries().size(), runs);
+                Integer[] queryOrder = PermutationUtils.preparePermutation(benchmark.getQueries().size(), finalThread);
+                List<QueryExecutionResult> queryExecutionResults = executeConcurrentQueries(benchmark, runs, reportStatus, lastTime, finalThread, queryOrder);
+                if (reportStatus && !queryExecutionResults.isEmpty()) {
+                    statusReporter.reportConcurrencyTestExecutionFinished(queryExecutionResults);
+                }
+                return queryExecutionResults;
+            });
+        }
+        return executionCallables;
+    }
 
-                        if (reportStatus) {
-                            statusReporter.reportExecutionFinished(result);
+    private List<QueryExecutionResult> executeConcurrentQueries(Benchmark benchmark, int runs, boolean reportStatus, Optional<ZonedDateTime> lastTime, int threadNumber, Integer[] queryOrder)
+            throws SQLException
+    {
+        boolean firstQuery = true;
+        List<QueryExecutionResult> queryExecutionResults = newArrayList();
+        try (Connection connection = getConnectionFor(new QueryExecution(benchmark, benchmark.getQueries().get(0), 0))) {
+            for (int run = 1; run <= runs; run++) {
+                for (int queryIndex = 0; queryIndex < benchmark.getQueries().size(); queryIndex++) {
+                    int permutedQueryIndex = queryIndex;
+                    if (!reportStatus) {
```

Comment:
don't we want to report status always?

Reply:
False reportStatus basically means we are doing pre-warming runs.